architecture/spmc.rst

```diff
@@ -0,0 +1,322 @@
+====
+SPMC
+====
+
+This document describes the SPMC(S-EL1) implementation for OP-TEE.
+More information on the SPMC can be found in the FF-A specification can be
+found in the
+`FF-A spec <https://developer.arm.com/documentation/den0077/latest>`_.
+
+.. toctree::
+ :numbered:
+
+SPMC Responsibilities
+=====================
+
+The SPMC is a critical component in the FF-A flow. Some of it mayor
+responsibilities are:
+
+- Initialisation and run-time management of the SPs:
+        The SPMC component is responsible for initialisation of the
+        Secure Partitions (Loading the Image, Setting up the stack, HEAP, ...).
+- Routing messages between endpoints:
+        The SPMC is responsible for passing FF-A messages from The Normal World
+        to SPs and back. It also needs to to pass FF-A messages between SPs
+- Memory management:
+        The SPMC is responsible for the memory management of the SPs. Memory
+        can be shared between SPs and from a SP to the Normal World.
+
+This document describes OP-TEE as a S-EL1 SPMC.
+
+Secure Partitions
+=================
+Secure Partitions (SPs) are the endpoints used in the FF-A protocol. When
+OP-TEE is used as a SPMC SPs run primarily inside S-EL0.
+
+The SPMC also exposes the OP-TEE kernel as an secure endpoint itself. This is
+used to handle all OP-TEE legacy operations. All legacy operations will be
+encapsulated inside FF-A messages. The OP-TEE endpoint will unpack the messages
+and afterwards handle them as standard OP-TEE calls. This is needed as
+TF-A (S-EL3) does only allow FF-A messages to be passed to the secure world
+when the SPMD is enabled.
```

Comment:
No, the only overhead is in how shared memory is registered. In the OP-TEE message model it can be a one-step process for temporary shared memory, while with FF-A it's always a two-step process (first register with SPMC and then invoke OP-TEE). It's covered at page 15 in https://static.linaro.org/connect/lvc21/presentations/lvc21-305.pdf

Reply:
OK, good. Thanks.